Implement HostPromiseRejectionTracker (#2530)

rhuanjl opened the pull request:
Responding to #2530 

This PR contains:
1. Internal CC machinery for HostPromiseRejectionTracker [ecmaspec 25.4.1.9](https://tc39.github.io/ecma262/#sec-host-promise-rejection-tracker) - this should enable any host to implement a handler for uncaught promise rejections e.g. the method defined here [WhatW spec 8.1.3.12](https://html.spec.whatwg.org/multipage/webappapis.html#unhandled-promise-rejections)
2. A very simplistic implementation in ch (behind a command line flag)
3. Test case

I'm expecting feedback/changes before this is merged but wanted to get other people's thoughts on it in this state. Please give me your thoughts/comments.

Cross ref: [Edge uservoice request for this feature](https://wpdev.uservoice.com/forums/257854-microsoft-edge-developer/suggestions/16665397-detect-unhandled-promise-rejection?tracking_code=4f4e218df62afae7db37fb5be5f4ff1c) - note this PR will not put the feature in edge - it just creates the hooks the edge development team would have to track them through.

**cc** @liminzhu @saschanaz @fatcerberus @MSLaguana 

**Notes:**
**1. Not spec compliant for "await" EDIT: snip - re-read spec this comment was wrong what I've done in this PR is correct**

**2. The ch implementation I've done is pretty bad,** it simply prints to the terminal whenever a notification is made, either handled or rejected - reasonable for testing that the interface is operational but certainly not what the WhatW spec would have a Host do.

**3. Parameters passed to the callback function** - per spec HostPromiseRejectionTracker should be passed the Promise and rejected/handled. I've gone for:
a) the promise
b) the promise's result - I know this could be retrieved from the promise object but I've added it for implementer's convenience
c) rejected/handled boolean - true for handled false for rejected - possibly should be an int or an enum of some kind

rhuanjl commented:
So all the windows test/debug builds are failing on the test I've written. Please could someone tell me how to make a baseline file that will pass on windows? (Reading the output shows that it's printing the right text)

My guess is that this is something to do with line endings  but I've tried setting my text editor to do CRLR line endings and it's still not working.

rhuanjl commented:
@boingoing thanks for the comments:
1. As far as I can CreateThenPromise is the only place we need SetIsHandled other than possibly an extra point for async functions. Both .then and .catch go through CreateThenPromise. await (for async) also goes through it but too late - see my longer comment above.

2. The third argument should probably change or at least be reversed - it's currently true when an unhandled rejection is notified and false for a "handled" notification - should perhaps be an int to allow for more options in the future - wasn't sure how to do an enum that would work from with JavascriptLibrary.cpp and also be visible from ChakraCore.h - hence not using a typedef enum for this.

3. I'll see what I can do about the line endings - I don't have a windows machine with any devtools of any kind on.

4. I've retargeted and rebased to master as requested

fatcerberus commented:
> I'm a little confused about the third argument (rejected) to the promise rejection tracker function. It looks like this indicates handled? All of the promises we pass to this function were rejected, yes?

To answer that question, yes, all promises passed to the callback are rejected.  The third parameter indicates whether the promise has just been rejected without a handler, or if it was rejected previously and is only now having a handler attached.  It corresponds to the `operation` parameter in the spec here:
https://tc39.github.io/ecma262/#sec-host-promise-rejection-tracker

liminzhu commented:
(only here to comment on the API surface change, not implementation)

Thanks for this very welcomed API addition @rhuanjl ! It looks nice and I appreciate it directly mirroring concepts in the spec. Couple of comments - 

1. @bterlson from the ecma spec perspective, how stable is https://tc39.github.io/ecma262/#sec-host-promise-rejection-tracker and the likelihood/risk it will be modified? I'm asking for guidance for @rhuanjl 's early comment of whether the `reject` bool in `JsHostPromiseRejectionTrackerCallback` (mirroring `operation` in `HostPromiseRejectionTracker`) should be as-is or changed to perhaps a var for reflexibility but more complexity to read the value.

2. @rhuanjl can you move the APIs to `ChakraCore.h`? ChakraCommon is for APIs living in both inbox Chakra and ChakraCore. We do want to put it in Chakra some time, but to put it there we'll have to go through a review process internally with the Windows API champions which will take some time, so it's a good idea to put them in ChakraCore.h first where all ChakraCore-only APIs lie.

rhuanjl commented:
@liminzhu thanks for the comments:
2. I've moved it offline, will test and ensure nothing breaks and re-push later
3. That is correct

fatcerberus commented on lib/Jsrt/ChakraCore.h:
```diff
@@ -130,6 +130,21 @@ typedef struct JsNativeFunctionInfo
+///     A Promise Rejection Tracker callback may be set - which will then be called when this occurs.
+/// </remarks>
+/// <param name="promise">The promise object, represented as a JsValueRef.</param>
+/// <param name="resolution">The value/cause of the rejection, represented as a JsValueRef.</param>
```
`resolution` should be renamed, to either `reason`--or if you want something more generic, `value`.  As far as promises go, "resolve" usually only refers to fulfillment (e.g. `Promise.resolve()` is never a rejection unless you specifically pass it a rejected promise) so the current name is maybe a bit confusing.

fatcerberus commented on lib/Jsrt/ChakraCore.h:
```diff
@@ -130,6 +130,21 @@ typedef struct JsNativeFunctionInfo
+/// </remarks>
+/// <param name="promise">The promise object, represented as a JsValueRef.</param>
+/// <param name="resolution">The value/cause of the rejection, represented as a JsValueRef.</param>
+/// <param name="reject">Boolean - true if the promise has just been rejected with no handler, 
```
As discussed offline I'm still of the opinion that if we keep this as a boolean, the meaning should be reversed, i.e. it should be false for "rejected" and true for "handler attached".

fatcerberus commented on lib/Jsrt/ChakraCore.h:
```diff
@@ -993,5 +1008,26 @@ CHAKRA_API
+///     Requires an active script context.
+///     </para>
+/// </remarks>
+/// <param name="callbackState"> The callback function being set.</param>
```
This should be
```cc
/// <param name="promiseRejectionTrackerCallback"> ... </param>
```
Correct?

rhuanjl commented:
@liminzhu I've adjusted for my friend @fatcerberus 's comments which seemed sensible to me which included:
1. changing resolution to reason - as the value given as that parameter is the reason the promise was rejected per spec.
2. swapping the 3rd parameter so that true indicates rejectionHandled and false indicates promiseRejected

I have also moved the API to chakracore.h

@akroshg do you know when you will have a chance to review this?

obastemur commented:
#4632 should fix it

jackhorton commented on lib/Runtime/Library/JavascriptLibrary.cpp:
```diff
@@ -5295,6 +5295,32 @@ namespace Js
+    void JavascriptLibrary::SetNativeHostPromiseRejectionTrackerCallback(HostPromiseRejectionTrackerCallback function)
+    {
+        this->nativeHostPromiseRejectionTracker = function;
+        this->hostShouldTrackPromiseRejections = true;
```
Nit: set `nativeHostPromiseRejectionTracker` to nullptr upfront so that you can avoid this bool and just check if the function pointer is nullptr or not.

rhuanjl replied:
Agreed - though I've now added a callbackState in the place of this.

jackhorton commented on lib/Runtime/Library/JavascriptPromise.h:
```diff
@@ -461,6 +461,8 @@ namespace Js
         };
 
+        bool GetIsHandled() {return isHandled;}
+        void SetIsHandled() {isHandled = true;}
```
Nit: spacing

rhuanjl replied:
updated

kfarnung commented on lib/Jsrt/ChakraCore.h:
```diff
@@ -993,5 +1008,26 @@ CHAKRA_API
         _In_ JsValueRef key,
         _Out_ bool *hasOwnProperty);
+
+
```
nit: extra newline

rhuanjl replied:
removed

kfarnung commented on lib/Runtime/Library/JavascriptLibrary.h:
```diff
@@ -949,6 +953,9 @@ namespace Js
         void SetNativeHostPromiseContinuationFunction(PromiseContinuationCallback function, void *state);
+        void SetNativeHostPromiseRejectionTrackerCallback(HostPromiseRejectionTrackerCallback function);
+        void CallNativeHostPromiseRejectionTracker(Var promise, Var reason, bool handled);
+
```
nit: extra newline

rhuanjl replied:
removed

kfarnung commented on bin/ch/ch.cpp:
```diff
@@ -757,6 +757,10 @@ HRESULT ExecuteTest(const char* fileName)
+        {
+            ChakraRTInterface::JsSetHostPromiseRejectionTracker(WScriptJsrt::PromiseRejectionTrackerCallback);
+        }
         len = strlen(fullPath);
```
nit: add a newline here

kfarnung replied:
Between the closing curly and the next line

rhuanjl replied:
added

kfarnung commented on bin/ch/WScriptJsrt.cpp:
```diff
@@ -1871,3 +1871,19 @@ void WScriptJsrt::PromiseContinuationCallback(JsValueRef task, void *callbackSta
+    Assert(reason != JS_INVALID_REFERENCE);
+    if(!handled)
+    {
+        wprintf(_u("Uncaught promise rejection\n"));
```
Should we also print out the message to ensure that the `reason` is being set correctly?

fatcerberus replied:
If I may add further commentary to this, it might make more sense to say "Promise rejected with no handler attached" rather than "Uncaught promise rejection".  We don't know whether or not it will remain uncaught at this time.

rhuanjl replied:
I've updated to print the reason.

I've left it as "Uncaught promise rejection" - as at the time of notification it is uncaught - it may be handled later but it is not at the moment of notification.

fatcerberus commented on bin/ChakraCore/ChakraCore.def:
```diff
@@ -60,3 +60,5 @@ JsLessThan
 
 JsCreateEnhancedFunction
+
+JsSetHostPromiseRejectionTracker
```
Should be a newline at end of file.

rhuanjl replied:
updated

fatcerberus commented on bin/ch/ChakraRtInterface.h:
```diff
@@ -356,6 +358,7 @@ class ChakraRTInterface
     static JsErrorCode WINAPI JsGetValueType(JsValueRef value, JsValueType *type) { return HOOK_JS_API(GetValueType(value, type)); }
     static JsErrorCode WINAPI JsSetIndexedProperty(JsValueRef object, JsValueRef index, JsValueRef value) { return HOOK_JS_API(SetIndexedProperty(object, index, value)); }
     static JsErrorCode WINAPI JsSetPromiseContinuationCallback(JsPromiseContinuationCallback callback, void *callbackState) { return HOOK_JS_API(SetPromiseContinuationCallback(callback, callbackState)); }
+    static JsErrorCode WINAPI JsSetHostPromiseRejectionTracker(JsHostPromiseRejectionTrackerCallback callback) { return HOOK_JS_API(SetHostPromiseRejectionTracker(callback)); }
```
Looking at the diff and seeing that literally all the other callbacks surrounding this get a `callbackState` argument, I think we should have it here for consistency too, even if I question its real-world utility in this case.

rhuanjl replied:
done

rhuanjl commented:
Thanks for the reviews @kfarnung @jackhorton @fatcerberus 
I think I've addressed everything. I haven't squashed yet in case anyone wants to check my new changes against their review points.

kfarnung commented on lib/Jsrt/Jsrt.cpp:
```diff
@@ -3457,6 +3457,17 @@ CHAKRA_API JsSetPromiseContinuationCallback(_In_opt_ JsPromiseContinuationCallba
+    /*allowInObjectBeforeCollectCallback*/true);
+}
+
+
```
nit: extra newlines

rhuanjl replied:
Oops, I'll fix this when I squash

fatcerberus commented on lib/Jsrt/ChakraCore.h:
```diff
@@ -993,5 +1008,26 @@ CHAKRA_API
+///     Requires an active script context.
+///     </para>
+/// </remarks>
+/// <param name="promiseRejectionTrackerCallback"> The callback function being set.</param>
```
Nit: Unnecessary leading space in description

rhuanjl commented:
Thanks again for the reviews - I've addressed all points and squashed.

Please let me know if there's anything else you want me to do with this.

sethbrenith commented on lib/Jsrt/ChakraCore.h:
```diff
@@ -130,6 +130,22 @@ typedef struct JsNativeFunctionInfo
+///     A Promise Rejection Tracker callback.
+/// </summary>
+/// <remarks>
+///     The host can specify a promise rejection tracker callback in <c>JsSetHostPromiseRejetionTracker</c>.
```
nit: missing `c` in `JsSetHostPromiseRejectionTracker`

rhuanjl replied:
updated

MSLaguana commented on test/es7/PromiseRejectionTracking.js:
```diff
@@ -0,0 +1,110 @@
+
+// Test HostPromiseRejectionTracker - see ecma262 section 25.4.1.9
+
+let tests = [
```
A couple of cases that I don't see captured here, and that I have some questions about:

```js
let controller;
let promise = new Promise((resolve, reject) => {  controller = {resolve, reject};});
let a = promise.then(() => {});
let b = promise.then(() => {});
controller.reject("Should see 2 unhandled rejections");

let c = a.then(() => {}); // Not clear: Should this have one "handled" and one "unhandled rejection" fire?

c.catch(() => {}); // Should have at least one handled fire, maybe 2 if the previous line did not fire a handled event?
```

fatcerberus replied:
@MSLaguana a.then() will fire one “handled” (for a) and one “rejected” (for c).  c.then() should fire another “handled” for c.

MSLaguana replied:
That makes sense. I think it'd be good to add these cases to the test file just to be sure as well.

rhuanjl replied:
I agree more cases may be a good idea - it gets quite messy working with the order of results due to the async nature of promises - the set I did was the most I could do without async results coming in in odd orders.

Note the above case gives just 1 notification:
As follows:
1. .reject() is called -> promise is handled (as the .then()s have occurred) - handling of .then()s scheduled for the next tick of the event loop (via promise continuation callback)
2. On reaching the event loop: a is handled by c. b -> rejected trigger-ring notification. c -> is handled by the .catch()

I can add some more cases before Monday (don't have much time today)

fatcerberus replied:
@rhuanjl is right, the example is completely synchronous and I forgot the .then()s got deferred till next tick.  Async code is confusing sometimes 🤪

MSLaguana replied:
Ah I see; and the tests that currently test the `handled` case only apply when the initial rejection immediately has no `then` and so the rejection notice is synchronous. I do think it'd be great if you could add a more complex test case in line with what I had above, but as you say I'm not sure how best to make it deterministic.

rhuanjl replied:
Wouldn’t be too bad if I did one file per case for complex cases.

I’m without a computer until Sunday now though so can’t do this today. Could make it a follow up PR.

rhuanjl replied:
I've now added some other more involved cases including the above.

MSLaguana commented on lib/Jsrt/ChakraCore.h:
```diff
@@ -130,6 +130,22 @@ typedef struct JsNativeFunctionInfo
+///     If a promise is rejected with no reactions or a reaction is added to a promise that was rejected
+///     before it had reactions by default nothing is done.
+///     A Promise Rejection Tracker callback may be set - which will then be called when this occurs.
+/// </remarks>
```
Please add a note here stating that this function must never throw any exceptions, as required by the spec.

I'd also like if you could add a comment about the lifetimes of the arguments; I believe that the promise handed to the callback here may be collected at any time (especially if it is an unhandled rejection which is never handled, or if it is a newly-handled rejection which will shortly fall out of scope and become unreferenced) and there are no particular steps taken to pin either of the arguments for this function.

fatcerberus replied:
I believe the specification includes wording implying the host is responsible for managing lifetime of the promise, but agreed it should be documented here too, for visibility if nothing else.

fatcerberus replied:
I note in particular that the spec specifically forbids the host from holding onto the promise after a “handled” event.

rhuanjl replied:
Updated comments and re-pushed

kfarnung commented on lib/Jsrt/Jsrt.cpp:
```diff
@@ -3452,6 +3452,15 @@ CHAKRA_API JsSetPromiseContinuationCallback(_In_opt_ JsPromiseContinuationCallba
     /*allowInObjectBeforeCollectCallback*/true);
 }
 
+CHAKRA_API JsSetHostPromiseRejectionTracker(_In_ JsHostPromiseRejectionTrackerCallback promiseRejectionTrackerCallback, _In_opt_ void *callbackState)
```
This needs to be moved into the `_CHAKRACOREBUILD` section at the end of the file.

kfarnung commented:
@curtisman @akroshg Can one of you take a look?

rhuanjl commented:
@kfarnung sorry I didn't respond on Friday, I was away without a computer.

I've rebased and squashed including your commit and adding some extra tests as requested by @MSLaguana

MSLaguana commented:
Thanks for adding those additional tests, looks good to me!

kfarnung commented on test/es7/PromiseRejectionTracking.js:
```diff
@@ -0,0 +1,145 @@
+
+let tests = [
+	{
+		name: "Reject promise with no reactions.",
```
super nit: looks like you used tabs here, can you switch over to spaces to match the majority of the files in the same directory?

rhuanjl replied:
Sorry about that, fixed.

rhuanjl commented:
@dotnet-bot test Windows arm_test please

rhuanjl commented:
@kfarnung any update on this?

kfarnung commented on bin/ch/WScriptJsrt.cpp:
```diff
@@ -1871,3 +1871,32 @@ void WScriptJsrt::PromiseContinuationCallback(JsValueRef task, void *callbackSta
+    JsValueRef strValue;
+    JsErrorCode error = ChakraRTInterface::JsConvertValueToString(reason, &strValue);
+
+    if(!handled)
```
nit: spacing between `if` and condition

kfarnung commented on bin/ch/ch.cpp:
```diff
@@ -757,6 +757,11 @@ HRESULT ExecuteTest(const char* fileName)
             IfFailGo(E_FAIL);
         }
 
+        if(HostConfigFlags::flags.TrackRejectedPromises)
```
nit: space between `if` and condition

kfarnung commented on lib/Runtime/Library/JavascriptLibrary.cpp:
```diff
@@ -5295,6 +5295,32 @@ namespace Js
+
+    void JavascriptLibrary::CallNativeHostPromiseRejectionTracker(Var promise, Var reason, bool handled)
+    {
+        if(nativeHostPromiseRejectionTracker != nullptr)
```
nit: space between `if` and condition

kfarnung replied:
nit: `this->nativeHostPromiseRejectionTracker`

kfarnung commented on lib/Runtime/Library/JavascriptLibrary.cpp:
```diff
@@ -5295,6 +5295,32 @@ namespace Js
+            BEGIN_LEAVE_SCRIPT(scriptContext);
+            try
+            {
+               nativeHostPromiseRejectionTracker(promise, reason, handled, this->nativeHostPromiseContinuationFunctionState);
```
nit: `this->nativeHostPromiseRejectionTracker`

kfarnung commented on lib/Runtime/Library/JavascriptPromise.cpp:
```diff
@@ -660,6 +661,10 @@ namespace Js
         {
             reactions = this->GetRejectReactions();
             newStatus = PromiseStatusCode_HasRejection;
+            if(!GetIsHandled())
```
nit: space between `if` and condition

kfarnung commented on lib/Runtime/Library/JavascriptPromise.cpp:
```diff
@@ -838,13 +843,19 @@ namespace Js
             EnqueuePromiseReactionTask(resolveReaction, sourcePromise->result, scriptContext);
             break;
         case PromiseStatusCode_HasRejection:
+            if(!sourcePromise->GetIsHandled())
```
nit: space between `if` and condition

kfarnung commented:
Hey @rhuanjl, thanks for your patience.  I'm still doing some local testing.  I pushed a minor commit to exclude the new test from one of our suites.

rhuanjl commented:
@kfarnung I fixed the new nits, I'll hold off on squashing for now as I assume there's more testing and at least one more review to come?

kfarnung commented:
@liminzhu @bterlson Any concerns with the API surface for this change?  I didn't see any follow up on the previous question:

> @bterlson from the ecma spec perspective, how stable is https://tc39.github.io/ecma262/#sec-host-promise-rejection-tracker and the likelihood/risk it will be modified? I'm asking for guidance for @rhuanjl 's early comment of whether the reject bool in JsHostPromiseRejectionTrackerCallback (mirroring operation in HostPromiseRejectionTracker) should be as-is or changed to perhaps a var for reflexibility but more complexity to read the value.

@rhuanjl I don't see any other issues with this change, I'm just trying to make sure I cover all the bases.

rhuanjl commented:
@kfarnung thanks very much for the help with this. Do you know if there is any chance of this getting landed soon.

I was hoping to prepare a PR for Promise.prototype.finally this weekend but it will be affecting several of the same files and nervous I'll end up with merge conflicts

kfarnung commented:
I'm still waiting for feedback from @liminzhu and @akroshg before landing this.  I don't have any concerns with the change and will land it as soon as I have approval from them.

akroshg commented on lib/Runtime/Library/JavascriptPromise.h:
```diff
@@ -470,6 +472,7 @@ namespace Js
     protected:
         Field(PromiseStatus) status;
         Field(Var) result;
+        Field(bool) isHandled;
```
nit: move this up above var just to get better packing? instead of 40 bytes it can be in 32 bytes.

rhuanjl replied:
Fixed

akroshg commented on test/es7/rlexe.xml:
```diff
@@ -92,4 +92,12 @@
+      <files>PromiseRejectionTracking.js</files>
+      <compile-flags>-TrackRejectedPromises -args summary -endargs -nodeferparse</compile-flags>
+      <baseline>PromiseRejectionTracking.baseline</baseline>
+      <tags>exclude_jshost</tags>
```
>exclude_jshost [](start = 12, length = 14)

Great. Thanks for covering with detailed test cases. @taylor.woll@microsoft.com should we be implementing this in jshost just to get consistency.

kfarnung replied:
I was the one who added the tag.  I think this is something we should do in the near future, it's just a matter of scheduling.

/cc @boingoing

akroshg commented:
@rhuanjl thanks for PR. Looks good to me. (apologize for adding a late comment).

liminzhu commented:
Looks good to me too!

rhuanjl commented:
Fixed last point and updated to latest.

Thanks for reviews everyone; sorry for the number of style issues, I'll try and be more consistent in future.

kfarnung commented:
@rhuanjl Thanks for your contribution!

liminzhu commented:
Thanks @rhuanjl ! Now that this has been merged, can I ask you to send a quick PR to https://github.com/Microsoft/ChakraCore-wiki to update the docs as well ([this](https://github.com/Microsoft/ChakraCore-wiki/commit/188d6b220cc5654abd2453dd84c8db1443bcff99) should be an example for what/where changes are needed). I'm happy to do that as well so let me know.
